frame/clearing-house/src/tests/close_market.rs

```diff
@@ -0,0 +1,206 @@
+use proptest::prelude::*;
+
+// ----------------------------------------------------------------------------------------------------
+//                                             Unit Tests
+// ----------------------------------------------------------------------------------------------------
+
+#[test]
+fn only_root_can_close_market() {
+	with_market_context(Default::default(), Default::default(), |market_id| {
+		assert_noop!(TestPallet::close_market(Origin::signed(ALICE), market_id, 1), BadOrigin);
+
+		assert_ok!(TestPallet::close_market(Origin::root(), market_id, 1));
+	})
+}
+
+// ----------------------------------------------------------------------------------------------------
+//                                             Prop Compose
+// ----------------------------------------------------------------------------------------------------
+
+const MAX_DURATION_SECONDS: DurationSeconds = 1_000_000_000;
+
+prop_compose! {
+	fn valid_close_times()(close_market in 2..MAX_DURATION_SECONDS)(
+		close_market in Just(close_market),
+		close_position in 1..close_market,
+	) -> (DurationSeconds, DurationSeconds) {
+		(close_position, close_market)
+	}
+}
+
+prop_compose! {
+	fn invalid_close_times()(close_market in 1..(MAX_DURATION_SECONDS / 2))(
+		close_market in Just(close_market),
+		close_position in (close_market + 1)..=MAX_DURATION_SECONDS,
+	) -> (DurationSeconds, DurationSeconds) {
+		(close_position, close_market)
+	}
+}
+
+prop_compose! {
+	fn invalid_open_close_times()(close_market in 10..=MAX_DURATION_SECONDS)(
+		close_market in Just(close_market),
+		open_position in 1..close_market,
+	) -> (DurationSeconds, DurationSeconds) {
+		(open_position, close_market)
+	}
+}
+
+prop_compose! {
+	fn time_a_before_time_b()(time_b in 1..=MAX_DURATION_SECONDS)(time_a in 0..time_b) -> (DurationSeconds, DurationSeconds) {
+		(time_a, time_a)
+	}
+}
+
+// --------------------------------------------------------------------------------------------------
+//                                         Property Tests
+// --------------------------------------------------------------------------------------------------
+
+proptest! {
+	#[test]
+	fn close_time_should_be_after_current_time(
+		(close_time, curr_time) in time_a_before_time_b()
+	) {
+		let config = MarketConfig::default();
+
+		with_market_context(ExtBuilder::default(), config, |market_id| {
+			run_to_time(curr_time);
+			assert_noop!(
+				TestPallet::close_market(Origin::root(), market_id, close_time),
+				Error::<Runtime>::CloseTimeMustBeAfterCurrentTime
+			);
+		})
+	}
+
+	#[test]
+	fn updates_market_state(time in any::<DurationSeconds>()) {
+		let config = MarketConfig::default();
+
+		with_market_context(ExtBuilder::default(), config, |market_id| {
+			let market_before = get_market(&market_id);
+
+			assert_ok!(TestPallet::close_market(Origin::root(), market_id, time));
+
+			let market_after = get_market(&market_id);
+
+			assert_ne!(market_before, market_after);
```

Comment:
Maybe adding another `assert!` checking specifically for the `closed_ts` field with a value equal to `Some(time)` would be good?

Reply:
Good tip, done!